docs/cartesian-product.adoc

```diff
@@ -300,37 +300,191 @@ To demonstrate with a table:
+This method must return `CartesianMethodArgumentsProvider.Sets`.
+`Sets` is a helper class, specifically for creating sets for `@CartesianTest`.
+To create the test data, instantiate a `new Sets()` then use the `add()` method to register the values for the parameters.
+When calling `add`, you have to pass in all the arguments you need for the corresponding parameter.
+For convenience, all `add()` methods return with your `Sets` instance, so you can chain `add()` calls.
+
+Let's look at an example.
+
+[source,java]
+----
+
+class MyCartesianTestClass {
+	@CartesianTest
+	@CartesianTest.Factory("setFactory")
+	void testMethod(String string, Class<?> clazz, TimeUnit unit) {
+		// passing test code
+	}
+
+	static Sets setFactory() {
+		return new Sets()
+			.add("Alpha", "Omega")
+			.add(Runnable.class, Cloneable.class, Predicate.class)
+			.add(TimeUnit.DAYS, TimeUnit.HOURS);
+	}
+}
+----
+
+The test `testMethod` is executed exactly twelve times.
+The first parameter can have any of the two values `"Alpha"` or `"Omega"`.
+The second parameter can have any of the three values `Runnable.class`, `Cloneable.class` or `Predicate.class`.
+The third parameter can have any of the two values `TimeUnit.DAYS` or `TimeUnit.HOURS`.
+`@CartesianTest` tests for all input combinations, that's `2 × 3 × 2`, so twelve tests in total.
+
+To demonstrate with a table:
+
+|===
+| # of test  | value of `string` | value of `clazz` | value of `unit`
+| 1st test   | "Alpha"           | Runnable.class   | TimeUnit.DAYS
+| 2nd test   | "Alpha"           | Runnable.class   | TimeUnit.HOURS
+| 3rd test   | "Alpha"           | Cloneable.class  | TimeUnit.DAYS
+| 4th test   | "Alpha"           | Cloneable.class  | TimeUnit.HOURS
+| 5th test   | "Alpha"           | Predicate.class  | TimeUnit.DAYS
+| 6th test   | "Alpha"           | Predicate.class  | TimeUnit.HOURS
+| 7th test   | "Omega"           | Runnable.class   | TimeUnit.DAYS
+| 8th test   | "Omega"           | Runnable.class   | TimeUnit.HOURS
+| 9th test   | "Omega"           | Cloneable.class  | TimeUnit.DAYS
+| 10th test  | "Omega"           | Cloneable.class  | TimeUnit.HOURS
+| 11th test  | "Omega"           | Predicate.class  | TimeUnit.DAYS
+| 12th test  | "Omega"           | Predicate.class  | TimeUnit.HOURS
+|===
+
+You can reuse the same argument provider method multiple times.
+
+[source,java]
+----
+class MyCartesianTestClass {
+	@CartesianTest
+	@CartesianTest.Factory("provideArguments")
+	void testNeedingArguments(String string, int i) {
+		// passing test code
+	}
+
+	@CartesianTest
+	@CartesianTest.Factory("provideArguments")
+	void testNeedingSameArguments(String string, int i) {
+		// different passing test code
+	}
+
+	static CartesianProductTest.Sets provideArguments() {
+		return new CartesianProductTest.Sets()
+			.add("Mercury", "Earth", "Venus")
+			.add(1, 12, 144);
+	}
```

Comment:
I don't understand why you don't piggy back on what is already available there and use `MethodSource`. I can write this like:

```
static Stream<Arguments> provideArguments() {
    return Stream.of(
        Arguments.of("Mercury", "Earth", "Venus"),
        Arguments.of(1, 12, 144),
    )
}
```

You will benefit of a lot of simplifications:

1. No need to do magic with reflection and you piggy back on the logic done by JUnit Jupiter (not sure if you can get an instance of the provider though)
2. You get nice auto completion and methods marked as used in IntelliJ

Reply:
First and foremost, because `Arguments` is not the right abstraction. [Its Javadoc](https://junit.org/junit5/docs/current/api/org.junit.jupiter.params/org/junit/jupiter/params/provider/Arguments.html) says:

> `Arguments` is an abstraction that provides access to an array of objects to be used for invoking a `@ParameterizedTest` method. 

What we need, though, is one "array of objects" per parameter (not per execution) and we're not invoking anything with them (they get folded into other "arrays", which then dictate the number of executions), let alone a `@ParameterizedTest`.

Re 1.: Unless we're reflectively accessing `MethodArgumentsProvider`, we still have to write our own reflective code to get these arguments. Even if we copy its source code, we'd still have to maintain it ourselves.

Re 2.: True, but I don't think that's worth the trade off.